swift/ql/lib/codeql/swift/regex/Regex.qll

```diff
@@ -0,0 +1,130 @@
+/**
+ * Provides classes and predicates for reasoning about regular expressions.
+ */
+
+import swift
+import codeql.swift.regex.RegexTreeView
+private import codeql.swift.dataflow.DataFlow
+private import internal.ParseRegex
+
+/**
+ * A data flow configuration for tracking string literals that are used as
+ * regular expressions.
+ */
+private module RegexUseConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node node) { node.asExpr() instanceof StringLiteralExpr }
+
+  predicate isSink(DataFlow::Node node) { node.asExpr() = any(RegexEval eval).getRegexInput() }
+
+  predicate isAdditionalFlowStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    // flow through `Regex` initializer, i.e. from a string to a `Regex` object.
+    exists(CallExpr call |
+      (
+        call.getStaticTarget().(Method).hasQualifiedName("Regex", ["init(_:)", "init(_:as:)"]) or
+        call.getStaticTarget()
+            .(Method)
+            .hasQualifiedName("NSRegularExpression", "init(pattern:options:)")
+      ) and
+      nodeFrom.asExpr() = call.getArgument(0).getExpr() and
+      nodeTo.asExpr() = call
+    )
+  }
+}
+
+private module RegexUseFlow = DataFlow::Global<RegexUseConfig>;
+
+/**
+ * A string literal that is used as a regular expression in a regular
+ * expression evaluation. For example the string literal `"(a|b).*"` in:
+ * ```
+ * Regex("(a|b).*").firstMatch(in: myString)
+ * ```
+ */
+private class ParsedStringRegex extends RegExp, StringLiteralExpr {
+  RegexEval eval;
+
+  ParsedStringRegex() {
+    RegexUseFlow::flow(DataFlow::exprNode(this), DataFlow::exprNode(eval.getRegexInput()))
+  }
+
+  /**
+   * Gets a call that evaluates this regular expression.
+   */
+  RegexEval getEval() { result = eval }
+}
+
+/**
+ * A call that evaluates a regular expression. For example, the call to `firstMatch` in:
+ * ```
+ * Regex("(a|b).*").firstMatch(in: myString)
+ * ```
+ */
+abstract class RegexEval extends CallExpr {
```

Comment:
Good question.  It isn't intended, but perhaps it should be.  Do you think anything should change about the design (or perhaps just the documentation) to reflect this?

Reply:
If we want to prevent users (and future us) from accidentally extending the set of regex evaluators we could replace
```ql
abstract class RegexEval extends CallExpr { ... }

/* ... */

private class AlwaysRegexEval extends RegexEval { ... }
```
with something like:
```ql
// This is now private to prevent users from accidentally extending the domain of the class.
private abstract class RegexEvalDomain extends CallExpr { ... }
// ... and we now expose only a final view of the class.
// This means that users writing `class MyRegexEval extends RegexEval { ... }` don't extend
// the domain of the class, but instead define a subclass (like they most likely intended).
final class RegexEval = RegexEvalDomain;

/* ... */

// And actual classes that need to extend the domain of the RegexEval class can do so by extending `RegexEvalDomain` (which you can only do in this file since the `RegexEvalDomain` class is private).
private class AlwaysRegexEval extends RegexEvalDomain { ... }
```

Note that "final extensions" is a very new QL feature so it won't be available until the next release. So we may want to wait with this change in order to not delay this PR.

Reply:
That's interesting, but to be honest I don't yet see much reason to prevent users from extending the class if they want to.

Perhaps the two variables shouldn't be exposed in the public interface though:
```
  Expr regexInput;
  Expr stringInput;
```
as that does lock us in to that design somewhat?

Reply:
> That's interesting, but to be honest I don't yet see much reason to prevent users from extending the class if they want to.

Yeah, it's certainly nice that it can be extended. We just need to ensure that none of our own queries in any suite extends this class as it invalidates (at the very least) the dataflow analysis that depends on the set of regex evaluators so that it'll be re-evaluated in the offending query.

> Perhaps the two variables shouldn't be exposed in the public interface though:
> ```
>  Expr regexInput;
>  Expr stringInput;
> ```
> as that does lock us in to that design somewhat?

That's true. Since the class is already abstract we could just require the presence of two abstract predicates `getRegexInput()` and `getStringInput()`.

Reply:
Done.